examples/5.MIS/main.jl

```diff
@@ -350,3 +350,280 @@ Bloqade.plot!(ax, Δ_final)
+prob = SchrodingerProblem(zero_state(nqubits(hamiltonian)), T_max, hamiltonian)
+emulate!(prob)
+
+# We can see that our algorithm found the Maximum Independent Set by summing the probabilities of being in any one of the MIS configurations:
+
+function get_MIS_prob(reg::Union{ArrayReg, SubspaceArrayReg}, configs) # want to maxmimize this 
+    prob = 0
+    x = [parse(Int, reverse(string(x)); base=2) for x in configs.c]
+    for (c, amp) in BloqadeMIS.ConfigAmplitude(reg)
+        if c in x
+            prob+=abs2(amp)
+        end
+    end
+    return prob
+end
+
+get_MIS_prob(prob.reg, configs)
+
+# ### Optimization
+
+# For even better results, we can optimize the pulse sequence for $\Omega$ and $\Delta$.
+# We can parametrize the pulses by $t_{start}$ and $t_{end}$, which are the times spent in the initial and final state, respectively. 
+# We will also optimize the shape on the detuning pulse by modeling it as a piecewise linear function given by parameters that indicate 
+# what values it reaches at regular time intervals. We will optimize for the probabilitity that of getting one the MIS solutions without 
+# applying any classical postprocessing on the results of the stimulation. 
+
+
+function get_waves(params::Vector{Float64}) # returns the pulses as a function of certain parameters
+    t_start, t_end, scale... = params
+    t_start+=0.03
+    T_max = 0.6
+    t_end = T_max - t_end - 0.03
+    t_interval = (t_end - t_start) / (length(scale) + 1)
+    Ω_max = 4 * 2π
+    Δ_start = -13 * 2π
+    Δ_end = Δ0 = 11 * 2π
+    
+    Δ_clock = [0.0, t_start]
+    Δ_val = [Δ_start, Δ_start]
+    for i in 1:length(scale)
+        push!(Δ_val, Δ0*scale[i])
+        push!(Δ_clock, t_start + i*t_interval)
+    end
+    push!(Δ_val, Δ_end); push!(Δ_val, Δ_end)
+    push!(Δ_clock, t_end); push!(Δ_clock, T_max)
+    Δ = piecewise_linear(clocks = Δ_clock, values = Δ_val);
+    Ω = piecewise_linear(clocks = [0.0, t_start, t_end, T_max], values = [0, Ω_max, Ω_max, 0]);
+    return Ω, Δ
+end
+
+# The example below plots the pulse sequence that would be generated by the parameters given in `x0`:
+
+function plot_waves(params::Vector{Float64})
+    Ω, Δ = get_waves(params)
+
+    graph, (ax1, ax2) = plt.subplots(ncols = 2, figsize = (12, 4))
+    Bloqade.plot!(ax1, Ω)
+    ax1.set_ylabel("Ω/2π (MHz)")
+    Bloqade.plot!(ax2, Δ)
+    ax2.set_ylabel("Δ/2π (MHz)")
+    return graph
+end
+
+x0 = [0.1, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8]
+
+plot_waves(x0)
+
+# We can now run the optimization using the `Nelder-Mead` algorithm
+
+function loss(atoms::AtomList, x::Vector{Float64}, configs)
+    Ω, Δ = get_waves(x)
+    hamiltonian = rydberg_h(atoms; Ω = Ω, Δ = Δ)
+    subspace = blockade_subspace(atoms, 4.5)
+    prob = SchrodingerProblem(zero_state(subspace), T_max, hamiltonian)
+    emulate!(prob)
+    return -get_MIS_prob(prob.reg, configs), prob.reg, Δ
+end
+
+optresult = Optim.optimize(x -> loss(atoms, x, configs)[1], x0)
```

Comment:
@johnzl-777 Why don't we use ForwardDiff.jl to do the optimization here?

Reply:
I think this is ok? It's general that should also work with hardware? or do we want to do a gradient-based optimization?